### ISSUE TITLE
Allow ArrayPool<T> to share pooled arrays of different types

### ISSUE THREAD

**Issue by kayle** [dotnet/coreclr#26638]
I didn't see any issues/discussion related to this...

It seems like a significant amount of effort has already gone into maximizing performance of `ArrayPool<T>`. How much additional effort would it be to allow 'free' array type conversions? Instead of `ArrayPool<byte>, ArrayPool<char>, ArrayPool<float>, ArrayPool<int>, ArrayPool<object>`, etc. all with their own buffer pools, just back everything from `ArrayPool<byte>`.

I think the main missing link is a method to change an array's type, or allocate a new array by stealing the buffer from another one. In terms of complexity, would it be similar to the new `GC.AllocateUninitializedArray<T>` api? 

Random implementation questions:
- Is it impossible to guarantee the old array is unreachable? (Unfortunately, I think yes)
- Is it possible to avoid catastrophic failures if code continued to use the old array? (maybe for unmanaged struct types, maybe not if a int[] field is now actually a byte[])
- Could an Array's data buffer be stored separate from the object header (e.g. allow creating array from Span or pointer/length)

**Comment by Emzi0767** [comment]
Why not simply rent a `byte[]` and create a Span<T> over it where T is an arbitrary value type?

**Comment by kayle** [comment]
I think it's the same reason that ArrayPool is used much more than MemoryPool - lots of existing code/public apis take array parameters: https://source.dot.net/#System.Private.CoreLib/shared/System/Buffers/ArrayPool.cs,704a680ba600a2a4

**Comment by GrabYourPitchforks** [comment]
Only certain array types are allowed to be cast to one another. For example, this is legal, supported, and fully type-safe:

```cs
int[] someArray = GetIntArray();
uint[] recastArray = (uint[])(object)someArray; // succeeds

Console.WriteLine(someArray.GetType()); // prints "int[]"
Console.WriteLine(recastArray.GetType()); // prints "int[]" (unexpected? perhaps, but 100% legal)
Console.WriteLine(ReferenceEquals(someArray, recastArray)); // prints "True"
```

This is only possible between primitive integral types of the same width: `byte <-> sbyte`, `short <-> ushort`, `int <-> uint`, and `long <-> ulong`. (For the purpose of allowed conversions, Enums are treated equivalently to their underlying integral type.)

In particular, it is __illegal__ to take a `T[]` and to cast it to a `U[]` where _T_ and _U_ have different widths. For example, there's no way to cast a `byte[]` to an `int[]`. The code below demonstrates one reason for this:

```cs
byte[] byteArray = new byte[128]; // 128 bytes in length
int[] intArray = reinterpret_cast<int[]>(byteArray);
Console.WriteLine(intArray.Length); // would print "128" (even though it should really read "32")
intArray[100] = 42; // corrupting arbitrary memory at byte offset 100 * sizeof(int) = byte offset 400
```

Additionally, the conversion is illegal when either `T` or `U` contains references. For example, conversion between `byte[]` and `object[]` is illegal. Even if you could solve the length problem demonstrated above, recall that the real actual backing array in memory has to be one of two types: `byte[]` or `object[]`. (This is what would be returned by `theArray.GetType()`.) This is critical because the .NET garbage collector needs to know whether the elements of the array represent binary data (as in the case of `byte[]`) or real object references that need to be followed and kept alive (`as in the case of object[]`). If the GC gets it wrong, it will cause memory corruption within the application. So the runtime forbids the application from casting a single array instance back and forth between these two worlds.

You had asked earlier if an array's buffer could be kept separate from the object header. Possibly. However, this could likely lead to an overall _increase_ in the number of objects on the managed heap, as the overwhelming majority of arrays are standalone and don't go through `ArrayPool<T>`, and all of those instances would now require two allocations. Additionally, this would have consequences at the p/invoke layers, since many applications rely on managed arrays having a very particular layout. Finally, it would potentially make `Object.ReferenceEquals` behave inconsistently, as two array references might compare as not equal but in actuality be backed by the same data buffer, which could cause application-level data corruption.

One potential action could be for the runtime to change the type of an array dynamically. For example, the runtime could overwrite the object header of an existing managed object in order to change its actual type. But this would be extraordinarily dangerous since it would need to be coordinated with the GC and could cause memory corruption issues if somebody improperly held on to the array instance after it was returned to the pool. So, while theoretically possible, I don't really see this happening.

If we're talking about projecting a `Span<T>` (instead of a `T[]`) over an existing `byte[]`, that's a more feasible proposition. _T_ would probably be restricted to only the built-in primitive integral types, and there might be some alignment concerns, but it's at least possible. But at the moment there's not significant demand for an API to rent / return such spans.

**Comment by stephentoub** [comment]
> If we're talking about projecting a Span<T> (instead of a T[]) over an existing byte[], that's a more feasible proposition

And already exists as MemoryMarshal.Cast.

**Comment by kayle** [comment]
Thanks for confirming the details. I agree that using `Span<T>` is correct where possible, and is probably the best approach for the future.

I realize this feedback is too late, but it seems like `ArrayPool<T>.Shared` should only be used for `byte` and reference types. By allowing separate instances for char/int, we're wasting memory.

Just scanning https://source.dot.net shows multiple violations. Let me know if there's an easy way to scan `ArrayPool<T>.Shared` references across a larger set of github repos.

https://source.dot.net/#System.Private.CoreLib/shared/System/Text/ValueStringBuilder.cs
Why not use `ArrayPool<byte>.Shared` and cast to `Span<char>`?

https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/ValueListBuilder.cs
This will use `ArrayPool<int>.Shared` based on current references, but could get worse in the future.

(I'm sure there was enough discussion on TextReader/TextWriter, but any reason not to obsolete/hide the array-based apis now that Spans are available? And isn't it currently backward calling the array-based api from the new span apis?)

**Comment by GrabYourPitchforks** [comment]
> Just scanning https://source.dot.net shows multiple violations.

I guess I don't see what you mean by "violation" in this statement. There's nothing wrong with the Framework's current use of `ArrayPool` - nothing that violates any best practice / security / reliability guidance we have.

There are also some practical considerations to modifying the current design. If we had an API that allowed you to rent `Span<T>` instances directly (regardless of what they're actually backed by), such a utility would also need a `Return(Span<T>)` API. But the conversion from managed object to `Span<T>` is one-way: you can't go backward and re-fetch the original `Array` or `string` reference from a `Span<T>`. So the `Return` method would have no way of knowing what actual object should be returned to the shared pool.

This could be worked around by having all callers use `ArrayPool<byte>.Shared.Rent` as you had suggested, then `MemoryMarshal.Cast` to the desired type. But that's now introducing unsafe-equivalent code into every single usage, which significantly harms the readability / reliability of the call sites.

We could instead return `CustomStruct<T>` which wraps `byte[]` and can project `Span<T>`, and the `Return` method could accept such structs, but there's a bit of usability and runtime weirdness with shuttling structs back and forth like this. It also harms readability of the call sites.

It's also sometimes simply not possible to project the desired `Span<T>` across a `byte[]`. For example, want a `Span<int>` whose length is 600,000,000 elements? No `byte[]` could ever be large enough to hold that much data. But using `int[]` as a backing store would work just fine.

In summary - the current design provides some nice benefits that we wouldn't really get with other designs. The current design isn't perceived as a shortcoming.

**Comment by kayle** [comment]
Yes, it's convenient to get T[] objects out. The primary goal of using ArrayPool is to improve memory/gc performance. Since `ArrayPool<T>.Shared` is already for perf optimized scenarios, convenience isn't necessarily the first priority. 

The problem is that `ArrayPool<T>.Shared` easily leads to increased WorkingSet and unnecessary 'leaks' in the form of stale cached buffers.

I propose adding guidance to use `ArrayPool<byte>.Shared` instead of  `ArrayPool<T>.Shared where T: unmanaged` wherever possible. Would you agree with this part?

The following shows the memory usage when using multiple ArrayPool instances:

```
public static void Main()
{
    Console.WriteLine("Allocate 2MB string");
    var longString = new string('a', 1024 * 1024);
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use string.Replace which depends on ArrayPool<int>.Shared");
    Parallel.For(0, 15, _ => longString.Replace("a", "b"));
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<int>.Shared directly, no memory increase since the pool is already populated");
    Parallel.For(0, 15, _ => UseArrays<int>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<byte>.Shared directly");
    Parallel.For(0, 15, _ => UseArrays<byte>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<char>.Shared directly");
    Parallel.For(0, 15, _ => UseArrays<char>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<uint>.Shared directly");
    Parallel.For(0, 15, _ => UseArrays<uint>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<float>.Shared directly");
    Parallel.For(0, 15, _ => UseArrays<float>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.WriteLine("use ArrayPool<double>.Shared directly");
    Parallel.For(0, 15, _ => UseArrays<double>());
    Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));

    Console.ReadLine();
}

private static void UseArrays<T>()
{
    for (int i = 0; i < 22; i++)
        ArrayPool<T>.Shared.Return(ArrayPool<T>.Shared.Rent(1 << i));
}
```

Here's the output on my machine:
```
Allocate 2MB string
2172232
use string.Replace which depends on ArrayPool<int>.Shared
44126592
use ArrayPool<int>.Shared directly, no memory increase since the pool is already populated
44132144
use ArrayPool<byte>.Shared directly
54621736
use ArrayPool<char>.Shared directly
75597008
use ArrayPool<uint>.Shared directly
117543760
use ArrayPool<float>.Shared directly
159490416
use ArrayPool<double>.Shared directly
260158056
```

**Comment by Sergio0694** [comment]
@kayle Small note, the following code should **never ever be used in production scenarios** as it could cause all sorts of issues, but since what you asked _is_ in fact technically possible, I thought I'd give it a try just for fun. Here's a method that lets you convert an array of a type to an array of _any other type_ that you want, as long as the input types are unmanaged types:

```csharp
/// <summary>
/// Morphs an array from a type to another
/// </summary>
/// <typeparam name="TFrom">The type of the input array</typeparam>
/// <typeparam name="TTo">The type of the resulting array</typeparam>
/// <param name="array">The input array to morph</param>
/// <returns>The input array as of type <typeparamref name="TTo"/></returns>
/// <remarks>Stop using the input array reference</remarks>
public static TTo[] Morph<TFrom, TTo>(this TFrom[] array)
    where TFrom : unmanaged
    where TTo : unmanaged
{
    if (array.Length == 0) return Array.Empty<TTo>();

    // Piggyback on Cast<TFrom, TTo> to calculate the length
    Span<TTo> span = MemoryMarshal.Cast<TFrom, TTo>(array);
    int length = span.Length;

    TTo[] dummy = new TTo[1];
    ref int r0 = ref Unsafe.As<TFrom, int>(ref array[0]);
    ref int r1 = ref Unsafe.As<TTo, int>(ref dummy[0]);

    // Overwrite method table and length
    Unsafe.Add(ref r0, -4) = Unsafe.Add(ref r1, -4);
    Unsafe.Add(ref r0, -2) = length;

    return (TTo[])(object)array;
}
```

The way this works is that as I'm sure you're well aware already, the `Array` class contains some additional fields before your actual data. In the case of a 1D array, going back from the data, it just contains the lower bound (0 in this case), then the length of the array, some other data and then the method table pointer (here's a nice [blog post](https://devblogs.microsoft.com/premier-developer/managed-object-internals-part-3-the-layout-of-a-managed-array-3/) about the memory layout of arrays). In that `Morph` method I'm first using `MemoryMarshal.Cast` just so that it calculates the appropriate length of the array of the new type for me, then I overwrite both the length field of the array (at offset -2 with respect to the array data) and then the method table pointer (at offset -4). This way when you do that final cast, the CLR thinks the array is of the right type and doesn't throw when you try to cast it.

With this trick, you could _in theory_ just use `ArrayPool<byte>` and then transform the returned `byte[]` arrays to the right type. You'd also probably need to revert their method table pointers before returning them though, otherwise the pool will probably not recognize them.

Again, just a proof of concept as an experiment, but it does work 😄

Here's an example of it in action

```csharp
float[] array1 = new float[4 * 128];
array1[0] = 3.14f;
array1[1] = 2;
array1[2] = 44;
array1[3] = 77;

Vector4[] array2 = array1.Morph<float, Vector4>();

Console.WriteLine(array2.GetType()); // System.Numerics.Vector4[]
Console.WriteLine(array2.Length); // 128
Console.WriteLine(array2[0]); // <3,14. 2. 44. 77>
```

_**NOTE:** doing this effectively breaks the type safety of the CLR, of course_

**Comment by GrabYourPitchforks** [comment]
> Again, just a proof of concept as an experiment, but it does work 😄

Until the garbage collector kicks in after you've overwritten the object header but before you've overwritten the length, blows past the end of the array, and starts corrupting other objects in the managed heap. 😉

**Comment by Sergio0694** [comment]
Well you're absolutely right, that's why I specifically mentioned twice not to ever, ever use stuff like this in actual production code 😄

**Comment by jkotas** [comment]
> you were to wrap those two assignments with GC.TryStartNoGCRegion

It does not work. `GC.TryStartNoGCRegion` is a best effort hint. It does not provide any hard guarantees. 

There is no way to do this reliably via exposed API. No amount of unsafe hacks is going to help. A long discussion about this in https://github.com/HelloKitty/Reinterpret.Net/issues/1 has examples that demonstrate the obscure runtime crashes you will get with this hack.

**Comment by Sergio0694** [comment]
@jkotas I see, thanks for clarifying that! And thank you for linking that related issue, it was a very interesting discussion to read and I definitely learnt something new! 😄

**Comment by danmosemsft** [comment]
I think this issue is resolved as not possible so I will close it.
